### PR TITLE
Allow Application.compile_env* to accept attributes

### PR DESCRIPTION
**Pull request by eksperimental**
In Elixir v1.12.0 it is giving the following error
when calling `Application.compile_env(@app, :key)`

    ** (FunctionClauseError) no function clause matching in Application."MACRO-compile_env"/4

### CODE REVIEW

**Comment by eksperimental** [comment]
Please apply to v1.13 and v1.12 branches.
Since we are deprecating Application.get_env/3 and friends, and Application.compile_env should be favored, I think it should be backported to v1.12 to promote wider adoption.

File: lib/mix/test/mix/tasks/archive_test.exs [hunk 1]
```diff
@@ -149,6 +149,9 @@ defmodule Mix.Tasks.ArchiveTest do
       true = Application.compile_env!(:git_repo, :archive_config)
 
       defmodule GitRepo.Archive do
+        @app :git_repo
+        true = Application.compile_env!(@app, :archive_config)
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/archive_test.exs:153]: Can you please add a test to `test/elixir/application_test.exs` instead? This test is not really related to the functionality.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
